### PR TITLE
Mapping improvements:

### PR DESCRIPTION
**Pull request by AsafMah**
- All mapping types now support all correct fields
- Better errors when mapping fails

### CODE REVIEW

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py [hunk 1]
```diff
@@ -65,18 +66,70 @@ class ColumnMapping:
...
-    # Json Mapping consts
     PATH = "Path"
     TRANSFORMATION_METHOD = "Transform"
-    # csv Mapping consts
     ORDINAL = "Ordinal"
     CONST_VALUE = "ConstValue"
-    # Avro Mapping consts
     FIELD_NAME = "Field"
+
+    # TODO - these are undocumented - do they really exist?
```
Comment (ohadbitt) [inline azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:75]: Columns is deprecated
StorageDataType seems like internal usage - delete both

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py [hunk 2]
```diff
@@ -65,18 +66,70 @@ class ColumnMapping:
...
+            IngestionMappingKind.SSTREAM,
+            IngestionMappingKind.PARQUET,
+            IngestionMappingKind.ORC,
+            IngestionMappingKind.W3CLOGFILE,
+        ],
+    }
+
+    NEEDED_PROPERTIES = invert_dict_of_lists(VALID_MAPPINGS)
+
+    CONSTANT_TRANSFORMATION_METHODS = [TransformationMethod.SOURCE_LOCATION.value, TransformationMethod.SOURCE_LINE_NUMBER.value]
```
Comment (ohadbitt) [inline azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:130]: the name doesnt exlpain why you choose these methods? are they common to all mappings?

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py [hunk 3]
```diff
@@ -107,21 +160,43 @@ def __init__(
...
+        for (prop, valid_types) in self.VALID_MAPPINGS.items():
+            if prop not in self.properties:
+                continue
+            if kind not in valid_types:
+                results.append(f"{prop} is not a valid mapping property for {kind}")
+
+        needed_props = self.NEEDED_PROPERTIES.get(kind, [])
+        if not needed_props:
+            results.append(f"{kind} is not a valid mapping kind")
+        else:
```
Comment (ohadbitt) [inline azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:178]: one line to elif?

**Review by ohadbitt — DISMISSED** [review]

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py [hunk 4]
```diff
@@ -107,21 +160,43 @@ def __init__(
...
+
+        for (prop, valid_types) in self.VALID_MAPPINGS.items():
+            if prop not in self.properties:
+                continue
+            if kind not in valid_types:
+                results.append(f"{prop} is not a valid mapping property for {kind}")
+
+        needed_props = self.NEEDED_PROPERTIES.get(kind, [])
+        if not needed_props:
+            results.append(f"{kind} is not a valid mapping kind")
```
Comment (ArielYehezkely) [inline azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:177]: This is a strange message - WDYM it's not a valid kind?

Comment (AsafMah) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:177]: It's not a key of the map, so it's not in IngestionMappingKind

If the user gives something else, it's a better error than the generic ones.

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py (hunk 3 above)
Comment (AsafMah) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:178]: The two ifs are ortagonal to each other

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py (hunk 2 above)
Comment (AsafMah) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:130]: These transformations replace the field entirely - meaning they don't need a `path` or a `field`, and if you use it it's wrong.

Other transformations are dependent on the `path` or `field` property, and if you use it without, it's wrong.

We actually had a similar check before the PR.

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py [hunk 5]
```diff
@@ -107,21 +160,43 @@ def __init__(
...
-            )
+            return ["Column name is required"]
+
+        results = []
+
+        for (prop, valid_types) in self.VALID_MAPPINGS.items():
+            if prop not in self.properties:
+                continue
+            if kind not in valid_types:
+                results.append(f"{prop} is not a valid mapping property for {kind}")
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:173]: Why not catch mistakes while we can?

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py (hunk 1 above)
Comment (AsafMah) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py]: Done

**Review by gunjandesai — DISMISSED** [review]

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py (hunk 4 above)
Comment (AsafMah) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py]: Ok removed

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py]: Ok  removed

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py [hunk 6]
```diff
@@ -107,21 +160,43 @@ def __init__(
         if storage_data_type:
             self.properties[self.STORAGE_DATA_TYPE] = storage_data_type
 
-    def is_valid(self, kind: IngestionMappingKind) -> bool:
+    def get_validation_errors(self, kind: IngestionMappingKind) -> List[str]:
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py]: Ok

File: azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py (hunk 2 above)
Comment (ohadbitt) [reply azure-kusto-ingest/azure/kusto/ingest/ingestion_properties.py:85]: can u add a comment for that

**Comment by gunjandesai** [comment]
signoff

**Pull request merged**